platforms/software/dependency-management/src/main/java/org/gradle/api/internal/attributes/ConcurrentIdentityCache.java

```diff
@@ -31,7 +31,17 @@ class ConcurrentIdentityCache<K, V> {
     private final Map<IdentityKey<K>, V> map = new ConcurrentHashMap<>();
 
     public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
-        return map.computeIfAbsent(new IdentityKey<>(key), k -> mappingFunction.apply(k.value));
+
+        IdentityKey<K> wrappedKey = new IdentityKey<>(key);
+
+        // First attempt to fetch the value from the map without locking
+        V result = map.get(wrappedKey);
```

Comment:
:x: My understanding is that `ConcurrentHashMap.computeIfAbsent` is already doing this: Check if the key is there and only acquire the lock if it isn't. So doing the extra check only adds overhead. Does this change show any improvements in our performance tests? 
<img width="799" alt="image" src="https://github.com/user-attachments/assets/847c98df-676c-45b6-99d8-4dcc26489248">

Reply:
❓ Indeed, just landed here from conversations on this topic and surprised as well.

I would say that only the change in `DefaultAttributesFactory` should have an impact.

Do we have a way to confirm this?

Reply:
The map only avoids acquiring the lock for the first node. If there are any hash collisions, which is increasing likely as the map fills up, we can end up chaining nodes for a single hash and therefore fallback to locking.

If you look at the implementation for `get`, there are no locks at all.

Reply:
The constructor _does_ provide a `concurrencyLevel` parameter, to be set to the number of threads performing updates, but that is only used to control the size of the initial table.

![image](https://github.com/user-attachments/assets/550d730b-8e7d-4705-a663-c958678c6cce)

Presumably, this is to ensure there are a number of nodes is greater than the number of threads performing updates, thus reducing the cases where there are collisions for a hash and reducing the need for synchronization in those cases

Reply:
In the case of the matcher cache (which is the most apparent in this graph), running the Android Extra Large 2000 project benchmark shows we get about _22_ entries in the cache by the end of the build (off the top of my head, from the last time I ran the benchmarks).

This cache gets hit a _lot_ -- once for every time we select a variant from a component, once for every time we perform artifact selection for a variant. And, it is dwarfed by reads. The hit rate on this cache is expected to be very high